__Cuaderno de trabajo de:__ Israel Carrera Manzaneque

## Lectura y escritura de archivos

    f = open("archivo.txt", "w")
    f.write('Hola mundo\n')
tenemos los siguientes modos:
- ‘r’ : Abre el archivo en modo lectura.
-‘w’ : write, escritura. Abre el archivo en modo escritura.
-‘a’ : append, se diferencia del modo ‘w’ en que en este caso no se sobreescribe el contenido del archivo, sino que se comienza a escribir al final del archivo.
-‘b’ : binary, binario.
-‘+’ : permite lectura y escritura simultáneas.
-‘U’ : universal newline, saltos de línea universales

In [1]:
f = open("archivo.txt", "r")
print(f)

<_io.TextIOWrapper name='archivo.txt' mode='r' encoding='UTF-8'>


In [2]:
f = open("archivo.txt", "r")
completo = f.read()
print(completo)
f.close()

1 2
2 4
3 6
4 8



In [3]:
f = open("archivo.txt", "r")
parte = f.read(8)
print(parte)
f.close()

1 2
2 4



In [15]:
f = open("archivo.txt", "r")
while True:
    linea = f.readline()
    if not linea: 
        break
    print(linea)

1 2

2 4

3 6

4 8



In [76]:
archivo="archivo.txt"
text=open(archivo).readlines()
nmaxlines=len(text)
for line in text:
    print(line.split())

['1', '2']
['2', '4']
['3', '6']
['4', '8']


## Manejando excepciones

In [6]:
archivo="archivo.txt"
try:
    f = open(archivo, 'r')
except IOError:
    print('no pude abrir', archivo)
else:
    print(archivo, 'tiene', len(f.readlines()), 'lineas')
    f.close()

archivo="no_existe.txt"
try:
    f = open(archivo, 'r')
except IOError:
    print('no pude abrir', archivo)
else:
    print(archivo, 'tiene', len(f.readlines()), 'lineas')
    f.close()

archivo.txt tiene 4 lineas
no pude abrir no_existe.txt


## Format
La forma más sencilla de mostrar algo en la salida estándar es median-
te el uso de la sentencia print, la cual permite también utilizar técnicas avanzadas de formateo, de forma similar al sprintf de C
Los especificadores más sencillos están formados por el símbolo % seguido de una letra que indica el tipo con el que formatear el valor Por ejemplo:

Especificador | Formato
:--------: | :-------:
%s |Cadena
%d |Entero
%o |Octal
%x |Hexadecimal
%f |Real


In [7]:
print ("%10s mundo" % "Hola")

      Hola mundo


In [8]:
print ("%-10s mundo" % "Hola")

Hola       mundo


In [24]:
from math import pi
print("%.4f" % pi)
print("%.6f" % pi)
print("%.4f" % pi,"%.6f" % pi)

3.1416
3.141593
3.1416 3.141593


In [10]:
print ("%.4s" % "hola mundo")

hola


In [68]:
n=123.456789
cadena='{0:12.6f}  '.format(n)
cadena=cadena+'{0:12.6f}  '.format(n)
cadena=cadena+"\n"
n=9876.54321
cadena=cadena+'{0:12.6f}  '.format(n)
cadena=cadena+'{0:12.6f}  '.format(n)
print(cadena)
print(type(cadena))

  123.456789    123.456789  
 9876.543210   9876.543210  
<class 'str'>


<hr>
<b><font color='red'>Ejercicio 01</font></b>  

Lee el archivo.txt 

In [48]:
f = open("archivo.txt", "r")
salida = f.read()
f.close()
final=""
for x in range(0,len(salida)):
    if x%2 == 0:
        if x%4 ==0:
            probando = salida[x]
            aux= int(salida[x])
            final = final + '{0:6.1f}  '.format(aux)
        else:
            probando = salida[x]
            aux= int(salida[x])
            final = final + '{0:6.2f}  '.format(aux) + "\n"
        
print(final)

   1.0    2.00  
   2.0    4.00  
   3.0    6.00  
   4.0    8.00  



y haz que tenga la siguiente salida:

    1.0   2.00  
    2.0   4.00  
    3.0   6.00  
    4.0   8.00  


<hr>
<b><font color='red'>Ejercicio 02</font></b>  

Haz lo mismo con el archvo archivo_error.txt,con ciddado de que a este archivo le falta algún dato, haz que en ese caso cuando lo lea, si no hay dos columnas no lo escriba.

In [49]:
archivo="archivo_error.txt"
t=open(archivo).readlines()
for line in t:
    aux = len(line.split())
    if(aux >= 2):
        print(line)

1 2

3 6

4 8



<hr>
<b><font color='red'>Ejercicio 03</font></b>  

Lee el archivo.txt, haz que imprima las columnas en las filas y las filas en las columnas

In [50]:
f = open("archivo.txt", "r")
salida = f.read()
f.close()
final=""
for x in range(0,len(salida)):
    if x%4 == 0:
        final = final + salida[x] + " "
final = final + "\n"
for x in range(2,len(salida)):
    if(x+2)%4 == 0:
        final = final + salida[x] + " "
        
print("Así es la salida inicial: ")
print(salida)
print("Así es la salida intercambiando filas y columnas: ")
print(final)

Así es la salida inicial: 
1 2
2 4
3 6
4 8

Así es la salida intercambiando filas y columnas: 
1 2 3 4 
2 4 6 8 


<hr>
<b><font color='red'>Ejercicio 04</font></b>  

1. Escribe en un archivo la siguiente matriz:
$
\begin{equation}
\begin{pmatrix}
3 & 1 & 3\\
-1 & 1 & 0\\
-2 & 4 & 1
\end{pmatrix}
\end{equation} 
$
2. Lee el archivo
3. Calcula con los datos leidos el determinante de la matriz y lo saque por pantalla

In [51]:
escribirMatriz = open("matriz.txt", "w")
escribirMatriz.write('3 1 3\n-1 1 0\n-2 4 1')
escribirMatriz.close()
f = open("matriz.txt", "r")
salida = f.read()
f.close()
matriz=salida.split()
suma1=int(matriz[0])*int(matriz[4])*int(matriz[8])
suma1=suma1+int(matriz[3])*int(matriz[7])*int(matriz[2])
suma1=suma1+int(matriz[6])*int(matriz[1])*int(matriz[5])

suma2=int(matriz[2])*int(matriz[4])*int(matriz[6])
suma2=suma2+int(matriz[5])*int(matriz[7])*int(matriz[0])
suma2=suma2+int(matriz[8])*int(matriz[1])*int(matriz[3])
determinante= suma1-suma2
print("El determinante de la matriz es: ", determinante)
# Entendiendo que lee todos los numeros, tenemos 9. Así pues vamos a multiplicar.

El determinante de la matriz es:  -2


<hr>
<b><font color='red'>Ejercicio 05</font></b>  

En el criptoanálisis, la técnica de análisis de frecuencias consiste en el aprovechamiento de estudios sobre la frecuencia de las letras o grupos de letras en los idiomas para poder establecer hipótesis para aprovecharlas para poder descifrar un texto cifrado sin tener la clave de descifrado (romper). Es un método típico para romper cifrados clásicos. 
Haz una anailisis de frecuencias para los 10 primeros càpitulos de el libro "el_quijote.txt"

In [52]:
abc="ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"
archivo = open("el_quijote.txt", "r")
quijote=archivo.read()
archivo.close()
quijote=quijote.upper()
quijote=quijote.replace("Á","A")
quijote=quijote.replace("É","E")
quijote=quijote.replace("Í","I")
quijote=quijote.replace("Ó","O")
quijote=quijote.replace("Ú","U")
longitud=0
for i in range (0, len(quijote)):
    if(abc.count(quijote[i])):
        longitud=longitud+1
        
        
for i in abc:
    print("%.2c  " % i,"%.3f"% (quijote.count(i)/longitud))
    #print("%.2c  " % i,"%.3f" % quijote[(abc.count(i))])

#for i in abc:
#    print("%.2c  " % i,"%.3f" % f[(abc.index(i))])

A   0.127
B   0.017
C   0.037
D   0.053
E   0.138
F   0.005
G   0.011
H   0.012
I   0.053
J   0.007
K   0.000
L   0.060
M   0.026
N   0.066
Ñ   0.000
O   0.099
P   0.021
Q   0.019
R   0.061
S   0.073
T   0.037
U   0.048
V   0.011
W   0.000
X   0.000
Y   0.015
Z   0.004


Salida:

    A   0.127
    B   0.017
    C   0.037
    D   0.053
    E   0.138
    F   0.005
    .
    .
    .

<hr>
<b><font color='red'>Ejercicio 06</font></b>  

haz un analisis de frecuencias en el libro encriptado "RH RQCEPESEFM AR MPM.txt", y comparala con el analisis anterior


In [53]:
abc="ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"
arch1="el_quijote.txt"
arch2="RH RQCEPESEFM AR MPM.txt"
archivo2=open(arch2,"r")
libro2=archivo2.read()
archivo1 = open(arch1, "r")
quijote=archivo1.read()
archivo2.close()
archivo1.close()
quijote=quijote.upper()
quijote=quijote.replace("Á","A")
quijote=quijote.replace("É","E")
quijote=quijote.replace("Í","I")
quijote=quijote.replace("Ó","O")
quijote=quijote.replace("Ú","U")
longitudQuijote=0
longitudLibro2=0
for i in range (0, len(quijote)):
    if(abc.count(quijote[i])):
        longitudQuijote=longitudQuijote+1
        
for i in range (0, len(libro2)):
    if(abc.count(libro2[i])):
        longitudLibro2=longitudLibro2+1

print(arch1, "   ", arch2)
for i in abc:
    print("%.2c  " % i,"%.3f"% (quijote.count(i)/longitudQuijote), "         %.2c  "% i, "%.3f"% (libro2.count(i)/longitudLibro2))
    
#print("el_quijote.txt","  RH RQCEPESEFM AR MPM.txt")    
#for i in abc:
#    print("%.2c  " % i,"%.3f" % f[(abc.index(i))],"       %.2c  " % i,"%.3f" % f2[(abc.index(i))])

el_quijote.txt     RH RQCEPESEFM AR MPM.txt
A   0.127          A   0.049
B   0.017          B   0.006
C   0.037          C   0.040
D   0.053          D   0.061
E   0.138          E   0.126
F   0.005          F   0.007
G   0.011          G   0.000
H   0.012          H   0.052
I   0.053          I   0.032
J   0.007          J   0.012
K   0.000          K   0.068
L   0.060          L   0.001
M   0.026          M   0.093
N   0.066          N   0.027
Ñ   0.000          Ñ   0.011
O   0.099          O   0.010
P   0.021          P   0.066
Q   0.019          Q   0.069
R   0.061          R   0.136
S   0.073          S   0.017
T   0.037          T   0.044
U   0.048          U   0.045
V   0.011          V   0.009
W   0.000          W   0.000
X   0.000          X   0.002
Y   0.015          Y   0.009
Z   0.004          Z   0.004


salida:

    el_quijote.txt   RH RQCEPESEFM AR MPM.txt
    A   0.127        A   0.049
    B   0.017        B   0.006
    C   0.037        C   0.040
    .                .
    .                .
    .                .

<hr>
<b><font color='red'>Ejercicio 07</font></b>  

Comparando el analisis de frecuencia sabrías decir cual es el titulo del libro.  
RH RQCEPESEFM AR MPM

In [54]:
abc="ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"
arch1="el_quijote.txt"
arch2="RH RQCEPESEFM AR MPM.txt"
archivo2=open(arch2,"r")
libro2=archivo2.read()
archivo1 = open(arch1, "r")
quijote=archivo1.read()
archivo2.close()
archivo1.close()
quijote=quijote.upper()
quijote=quijote.replace("Á","A")
quijote=quijote.replace("É","E")
quijote=quijote.replace("Í","I")
quijote=quijote.replace("Ó","O")
quijote=quijote.replace("Ú","U")
longitudQuijote=0
longitudLibro2=0
for i in range (0, len(quijote)):
    if(abc.count(quijote[i])):
        longitudQuijote=longitudQuijote+1
        
for i in range (0, len(libro2)):
    if(abc.count(libro2[i])):
        longitudLibro2=longitudLibro2+1

analisisQuijote = { }
analisisSegundoLibro = { }

for i in abc:
    analisisQuijote.update({'{0:6.3f}'.format(quijote.count(i)/longitudQuijote):i})
    analisisSegundoLibro.update({'{0:6.3f}'.format(libro2.count(i)/longitudLibro2):i})



analisisQuijote=sorted(analisisQuijote.items(), reverse=True)
analisisSegundoLibro=sorted(analisisSegundoLibro.items(), reverse=True)

for i in range(0,len(analisisQuijote)):
               print(analisisQuijote[i], "       ", analisisSegundoLibro[i])

# la traducción:
# RH RQCEPESEFM AR MPM = ED ESCARAQAJO IE ORO
#Qué sería "EL ESCARABAJO DE ORO" de Allan Poe.

(' 0.138', 'E')         (' 0.136', 'R')
(' 0.127', 'A')         (' 0.126', 'E')
(' 0.099', 'O')         (' 0.093', 'M')
(' 0.073', 'S')         (' 0.069', 'Q')
(' 0.066', 'N')         (' 0.068', 'K')
(' 0.061', 'R')         (' 0.066', 'P')
(' 0.060', 'L')         (' 0.061', 'D')
(' 0.053', 'I')         (' 0.052', 'H')
(' 0.048', 'U')         (' 0.049', 'A')
(' 0.037', 'T')         (' 0.045', 'U')
(' 0.026', 'M')         (' 0.044', 'T')
(' 0.021', 'P')         (' 0.040', 'C')
(' 0.019', 'Q')         (' 0.032', 'I')
(' 0.017', 'B')         (' 0.027', 'N')
(' 0.015', 'Y')         (' 0.017', 'S')
(' 0.012', 'H')         (' 0.012', 'J')
(' 0.011', 'V')         (' 0.011', 'Ñ')
(' 0.007', 'J')         (' 0.010', 'O')
(' 0.005', 'F')         (' 0.009', 'Y')
(' 0.004', 'Z')         (' 0.007', 'F')
(' 0.000', 'X')         (' 0.006', 'B')
